### PR TITLE
MFEM Code of Conduct

### PR DESCRIPTION
**Pull request by tzanio**
Adding a code of conduct document based on the [Contributor Covenant][homepage], version 2.0.

[homepage]: https://www.contributor-covenant.org

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2702](https://github.com/mfem/mfem/pull/2702) | @tzanio | @tzanio | @acfisher + @pazner | 12/10/21 | 12/16/21 | 12/16/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+Examples of representing our community include using an official e-mail address,
+posting via an official social media account, or acting as an appointed
+representative at an online or offline event.
+
+## Enforcement
+
+Instances of abusive, harassing, or otherwise unacceptable behavior may be
+reported to the community leaders responsible for enforcement at
+mfem@llnl.gov.
+All complaints will be reviewed and investigated promptly and fairly.
```
Comment (acfisher) [inline CODE_OF_CONDUCT.md:64]: How about we add the following here:
Any leader involved in the reported behavior will recuse themselves from the investigation and decision making about the resolution of the complaint.

Comment (tzanio) [reply CODE_OF_CONDUCT.md:64]: We can either say "Anyone" instead of "Any community leader"

**Review by pazner — APPROVED** [review]

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (tzanio) [reply CODE_OF_CONDUCT.md:63]: Addressed.

[the remark this answers is not in the document]
Comment (tzanio) [reply CODE_OF_CONDUCT.md]: Addressed.

**Comment by tzanio** [comment]
@mfem/developers -- this PR is a proposed code of conduct for the project based on the popular "Contributor Covenant".

Please read it carefully and speak up if you have any questions or concerns.

I'd like to get explicit approval on this PR from all MFEM developers as an affirmative statement of intent to abide by and enforce the code of conduct.

**Review by camierjs — APPROVED** [review]

**Review by jamiebramwell — APPROVED** [review]

**Review by artv3 — APPROVED** [review]

**Review by cjvogl — APPROVED** [review]

**Review by brendankeith — APPROVED** [review]

**Review by mlstowell — APPROVED** [review]

**Review by YohannDudouit — APPROVED** [review]

**Review by kmittal2 — APPROVED** [review]

**Review by rcarson3 — APPROVED** [review]

**Review by bslazarov — APPROVED** [review]

**Review by vladotomov — APPROVED** [review]

**Review by dylan-copeland — APPROVED** [review]

**Review by psocratis — APPROVED** [review]

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -21,6 +21,8 @@ documentation; new examples and miniapps; HPC performance improvements; etc.
 MFEM is distributed under the terms of the BSD-3 license. All new contributions
 must be made under this license.
 
+Note also that MFEM has a [Code of Conduct](CODE_OF_CONDUCT.md). By participating in the MFEM community, you agree to abide by its rules.
```
Comment (barker29) [inline CONTRIBUTING.md:24]: As a very minor comment, this line does not match the linewidth of this and other `.md` files.

**Review by barker29 — APPROVED** [review]

File: CONTRIBUTING.md (hunk 2 above)
Comment (tzanio) [reply CONTRIBUTING.md]: Fixed :)

**Review by jakubcerveny — APPROVED** [review]

**Comment by v-dobrev** [comment]
> Adding a code of conduct document based on the [Contributor Covenant](https://www.contributor-covenant.org), version 2.0.

I see version 2.1 on the website. Did you use version 2.0 on purpose? Or was 2.0 the latest version when you started this PR?

File: CODE_OF_CONDUCT.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+**Consequence**: A temporary ban from any sort of interaction or public
+communication with the community for a specified period of time. No public or
+private interaction with the people involved, including unsolicited interaction
+with those enforcing the Code of Conduct, is allowed during this period.
+Violating these terms may lead to a permanent ban.
+
+### 4. Permanent Ban
+
+**Community Impact**: Demonstrating a pattern of violation of community
+standards, including sustained inappropriate behavior,  harassment of an
```
Comment (v-dobrev) [inline CODE_OF_CONDUCT.md:110]:
```suggestion
standards, including sustained inappropriate behavior, harassment of an
```

File: CODE_OF_CONDUCT.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
+# MFEM Code of Conduct
+
+## Our Pledge
+
+We as members, contributors, and leaders pledge to make participation in our
+community a harassment-free experience for everyone, regardless of age, body
+size, visible or invisible disability, ethnicity, sex characteristics, gender
+identity and expression, level of experience, education, socio-economic status,
+nationality, personal appearance, race, religion, or sexual identity
+and orientation.
```
Comment (v-dobrev) [inline CODE_OF_CONDUCT.md:10]:
```suggestion
nationality, personal appearance, race, religion, or sexual identity and
orientation.
```

File: CODE_OF_CONDUCT.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+Examples of behavior that contributes to a positive environment for our
+community include:
+
+* Demonstrating empathy and kindness toward other people
+* Being respectful of differing opinions, viewpoints, and experiences
+* Giving and gracefully accepting constructive feedback
+* Accepting responsibility and apologizing to those affected by our mistakes,
+  and learning from the experience
+* Focusing on what is best not just for us as individuals, but for the
+  overall community
```
Comment (v-dobrev) [inline CODE_OF_CONDUCT.md:26]:
```suggestion
* Focusing on what is best not just for us as individuals, but for the overall
  community
```

File: CODE_OF_CONDUCT.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+Examples of unacceptable behavior include:
+
+* The use of sexualized language or imagery, and sexual attention or
+  advances of any kind
```
Comment (v-dobrev) [inline CODE_OF_CONDUCT.md:31]:
```suggestion
* The use of sexualized language or imagery, and sexual attention or advances of
  any kind
```

File: CODE_OF_CONDUCT.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+* Trolling, insulting or derogatory comments, and personal or political attacks
+* Public or private harassment
+* Publishing others' private information, such as a physical or email
+  address, without their explicit permission
```
Comment (v-dobrev) [inline CODE_OF_CONDUCT.md:35]:
```suggestion
* Publishing others' private information, such as a physical or email address,
  without their explicit permission
```

File: CODE_OF_CONDUCT.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+unprofessional or unwelcome in the community.
+
+**Consequence**: A private, written warning from community leaders, providing
+clarity around the nature of the violation and an explanation of why the
+behavior was inappropriate. A public apology may be requested.
+
+### 2. Warning
+
+**Community Impact**: A violation through a single incident or series
+of actions.
```
Comment (v-dobrev) [inline CODE_OF_CONDUCT.md:87]:
```suggestion
**Community Impact**: A violation through a single incident or series of
actions.
```

File: CODE_OF_CONDUCT.md [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+**Consequence**: A warning with consequences for continued behavior. No
+interaction with the people involved, including unsolicited interaction with
+those enforcing the Code of Conduct, for a specified period of time. This
+includes avoiding interactions in community spaces as well as external channels
+like social media. Violating these terms may lead to a temporary or
+permanent ban.
```
Comment (v-dobrev) [inline CODE_OF_CONDUCT.md:94]:
```suggestion
like social media. Violating these terms may lead to a temporary or permanent
ban.
```

File: CODE_OF_CONDUCT.md [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+Violating these terms may lead to a permanent ban.
+
+### 4. Permanent Ban
+
+**Community Impact**: Demonstrating a pattern of violation of community
+standards, including sustained inappropriate behavior,  harassment of an
+individual, or aggression toward or disparagement of classes of individuals.
+
+**Consequence**: A permanent ban from any sort of public interaction within
+the community.
```
Comment (v-dobrev) [inline CODE_OF_CONDUCT.md:114]:
```suggestion
**Consequence**: A permanent ban from any sort of public interaction within the
community.
```

**Comment by v-dobrev** [comment]
> > Adding a code of conduct document based on the [Contributor Covenant](https://www.contributor-covenant.org), version 2.0.
> 
> I see version 2.1 on the website. Did you use version 2.0 on purpose? Or was 2.0 the latest version when you started this PR?

There are a few small changes in v2.1 -- we can add them and update the attribution references to v2.1.

**Comment by tzanio** [comment]
> I see version 2.1 on the website. Did you use version 2.0 on purpose? Or was 2.0 the latest version when you started this PR?

This is the version that GitHub suggested at https://github.com/mfem/mfem/community when I started the PR.

**Review by rw-anderson — APPROVED** [review]

**Comment by tzanio** [comment]
Updated to v2.1. The changes were minimal.

**Review by v-dobrev — APPROVED** [review]

File: CONTRIBUTING.md (hunk 2 above)
Comment (acfisher) [reply CONTRIBUTING.md]: Great to see you are still around @barker29

**Review by acfisher — APPROVED** [review]

**Review by gillette7 — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**